### PR TITLE
More optimize dlk runtime

### PR DESCRIPTION
**Pull request by primenumber**
| |DE10-Nano (CPU)|DE10-Nano (FPGA)|Raspberry Pi 3B|
|----|----|----|----|
|before|327ms|28.7ms|528ms|
|by #353 |129ms|28.7ms|83ms|
|by #359 |94ms|19.9ms|52ms|
|by #381 |94ms|17.7ms|52ms|
|after|61.3ms|13.3ms|36.2ms|

| |Core i7-8750H@governor=performance|Xeon Gold 6154|
|----|----|----|
|before|49.8ms|100ms|
|by #365 |2.34ms@OMP_NUM_THREADS=6|7.43ms@OMP_NUM_THREADS=2|
|by #373 |1.73ms@OMP_NUM_THREADS=6|1.85ms@OMP_NUM_THREADS=18|
|after|1.42ms@OMP_NUM_THREADS=6|1.36ms@OMP_NUM_THREADS=18|

Execution time varies, these are minimum time.

## Motivation and Context

This PR implements a lot of generic or target specific optimizations. 

## Description

Optimize these elements:
- float matrix multiplication
- quantizer
- pack_input
- lookup
- extract_image_patches
- convert_tensor
- memcpy
- copy thresholds
- conv2d+apply_thresholds+pack_16bit+convert_tensor

## How has this been tested?

```shell
$ PYTHONPATH=python/dlk python python/dlk/scripts/generate_project.py -i minimal_graph_with_shape.pb -o tmp/ -p object_detection -hq -ts -cache
$ cd tmp/object_detection.prj
$ CXXFLAGS=-DFUNC_TIME_MEASUREMENT make -j8 lm_x86_avx
$ make clean
$ CXXFLAGS=-DFUNC_TIME_MEASUREMENT make -j8 lm_fpga
$ make clean
$ CXXFLAGS=-DFUNC_TIME_MEASUREMENT make -j8 lm_arm
$ make clean
$ CXXFLAGS=-DFUNC_TIME_MEASUREMENT make -j8 lm_aarch64
$ make clean
$ CXXFLAGS=-DFUNC_TIME_MEASUREMENT make -j8 lm_x86
```

and test each executable file.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
run dlk test

**Comment by iizukak** [comment]
@primenumber some dlk test failed. please check

**Comment by tkng** [comment]
Some tests doesn't passed, something wrong happens, but I cannot understand how to execute unit test on my local pc... Please give me 3 days to understand.

**Comment by primenumber** [comment]
Runtime may be terminated due to SIGSEGV...

**Comment by primenumber** [comment]
I found large array allocated on stack area, it caused stack overflow.
I fixed it.

**Comment by iizukak** [comment]
@tkng how about your review progress.

**Comment by primenumber** [comment]
run test

File: dlk/python/dlk/templates/include/de10_nano.h [hunk 1]
```diff
@@ -505,28 +505,22 @@ Parameters calcParameters(uint32_t inputHeight, uint32_t inputWidth, uint32_t in
 }
 
 void RunTCA(unsigned long input_addr, unsigned long output_addr, unsigned long kernel_addr,
-  BIN_CONV_OUTPUT th_data[], unsigned in_w, unsigned in_h, unsigned in_c, unsigned nbits_in_data,
+  unsigned long thresholds_addr, unsigned in_w, unsigned in_h, unsigned in_c, unsigned nbits_in_data,
   unsigned out_w, unsigned out_h, unsigned out_c, unsigned k_w, unsigned k_h, unsigned pad, unsigned stride) {
 
   const unsigned k_size = (k_h * k_w * in_c * out_c) / 32;
   // MappedMem k_data_mem(KERNEL_ADDR, k_size, sizeof(T_UINT));
   // k_data_mem.Write(k_data_packed, k_size);
```
Comment (tkng) [inline dlk/python/dlk/templates/include/de10_nano.h]: These commented lines should be kept, or better to remove?

File: dlk/python/dlk/templates/include/quantizer.h [hunk 2]
```diff
@@ -85,118 +46,16 @@ void func_QTZ_binary_mean_scaling(
...
-  float32x4_t min_value_x4 = vdupq_n_f32(min_value);
-  float32x4_t round_offset = vdupq_n_f32(0.5);
-  float32x4_t max_value_rn = vdupq_n_f32(max_value_r * n);
-
-  for (; i <= static_cast<int>(end) - 4; i += 4)
-  {
-    float32x4_t tmp = vld1q_f32(&input[i]);
-    tmp = vmaxq_f32(tmp, min_value_x4);
-    tmp = vminq_f32(tmp, max_value_x4);
-    //    tmp = vmlaq_f32(tmp, max_value_rn, round_offset);
```
Comment (tkng) [inline dlk/python/dlk/templates/include/quantizer.h]: Seems that it's better to be removed.

File: dlk/python/dlk/templates/include/quantizer.h [hunk 3]
```diff
@@ -85,118 +46,16 @@ void func_QTZ_binary_mean_scaling(
...
-                                              std::min(i + chunk_size, static_cast<unsigned int>(num_elems)));
-    }));
-  }
-
-  for (auto& th: threads) {
-    th.join();
-  }
-
-  //static T_UINT counter = 0;
-  //write_to_file("out/qconv_input_quantized_not_packed", counter++, output_not_packed, in_height * in_width * in_depth);
```
Comment (tkng) [inline dlk/python/dlk/templates/include/quantizer.h]: These commented lines looks better to be removed.

File: dlk/python/dlk/templates/src/quantizer.cpp [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  for (; i <= static_cast<int>(end) - 8; i += 8)
+  {
+    const auto in0 = vld1q_f32(&input[i]);
+    const auto in1 = vld1q_f32(&input[i + 4]);
+    const auto mx0 = vmaxq_f32(in0, min_value_x4);
+    const auto mx1 = vmaxq_f32(in1, min_value_x4);
+    const auto mn0 = vminq_f32(mx0, max_value_x4);
+    const auto mn1 = vminq_f32(mx1, max_value_x4);
+    //const auto mad0 = vmlaq_f32(round_offset, mn0, max_value_rn);
+    //const auto mad1 = vmlaq_f32(round_offset, mn1, max_value_rn);
```
Comment (tkng) [inline dlk/python/dlk/templates/src/quantizer.cpp]: Looks that it's better to be removed

File: dlk/python/dlk/templates/include/quantizer.h (hunk 2 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/quantizer.h]: here is deleted.

File: dlk/python/dlk/templates/include/quantizer.h (hunk 3 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/quantizer.h]: here is deleted.

File: dlk/python/dlk/templates/include/de10_nano.h (hunk 1 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/de10_nano.h]: Removed.

File: dlk/python/dlk/templates/src/quantizer.cpp (hunk 4 above)
Comment (primenumber) [reply dlk/python/dlk/templates/src/quantizer.cpp]: Removed.

**Comment by tkng** [comment]
- [x] Readabilty approval
- [x] Ownership approval

let's merge after unit tests are passed

**Pull request merged**